File: 1_theory_tools/exercises/work/Exercise_3.py
# import packages used
import numpy as np
import scipy.optimize as optimize
from scipy import interpolate # Interpolation routines

def solve_consumption_grid_search(par):
     # initialize solution class
    class sol: pass
    sol.C = np.zeros(par.num_W)
    sol.V = np.zeros(par.num_W)
    
    # consumption grid as a share of available resources
    grid_C = np.linspace(0.0,1.0,par.num_C) 
    
    # Resource grid
    grid_W = par.grid_W

    # Init for VFI
    sol.delta = 1000 #difference between V_next and V_now
    sol.it = 0  #iteration counter 
    
    while (par.max_iter>= sol.it and par.tol<sol.delta):
        sol.it +=1
        V_next = sol.V.copy()
        interp = interpolate.interp1d(par.grid_W,V_next,bounds_error=False,fill_value='extrapolate')
        for iw,w in enumerate(grid_W):  # enumerate automaticcaly unpack w
            fun = lambda x: -V(x,w,interp,par)       
            
            res = optimize.minimize(fun,0.5,bounds =((1e-8,1-1e-8),))
            
            sol.V[iw] = -res.fun
            sol.C[iw] = res.x*w
            
            
        sol.delta = np.amax(np.abs(sol.V - V_next))
    
    return sol


def solve_consumption_grid_search2(par):
    
     # initialize solution class
    class sol: pass
    sol.C = np.zeros(par.num_W)
    sol.V = np.zeros(par.num_W)
    
    # consumption grid as a share of available resources
    grid_C = np.linspace(0.0,1.0,par.num_C) 
    
    # Resource grid
    grid_W = par.grid_W

    # Init for VFI
    sol.delta = 1000 #difference between V_next and V_now
    sol.it = 0  #iteration counter 
    
    while (par.max_iter>= sol.it and par.tol<sol.delta):
        sol.it +=1
        V_next = sol.V.copy()
        for iw,w in enumerate(grid_W):  # enumerate automaticcaly unpack w
            c = grid_C*w

            wt1 = w-c
            V_guess = np.sqrt(c)+par.beta * np.interp(wt1,grid_W,V_next)

            i = np.argmax(V_guess)
            

            sol.V[iw] = V_guess[i]
            sol.C[iw] = c[i]
            
            
        sol.delta = np.amax(np.abs(sol.V - V_next))
    
    return sol


def V(x,w,interp,par):
    #"unpack" c
    if type(x) == np.ndarray: # vector-type: depends on the type of solver used
        c = w*x[0] 
    else:
        c = w*x

    wt1 = w-c

    return np.sqrt(c)+par.beta*interp(wt1)

File: 1_theory_tools/exercises/Julia/Exercise_3.jl


module ex3
using Interpolations
using Optim
using Parameters # For setting default values in struct with @with_kw
using UnPack # For unpacking 

@with_kw mutable struct Sol
    W_grid::Array
    C::Array
    V::Array
    delta::Float64 = 100000.
    it::Int64 = 0
end

function solve_consumption_grid_search(par::NamedTuple)
    
    
    sol = Sol(W_grid = par.W_grid,
            C= zeros(Float64,par.num_W),
            V= zeros(Float64,par.num_W),
            )
    
    while (par.max_iter>= sol.it) & (par.tol<sol.delta)
        sol.it +=1
        V_next = copy(sol.V)

        
        interp = LinearInterpolation(par.W_grid,V_next)

        for (iw,w) in enumerate(par.W_grid)  # enumerate automaticcaly unpack w
            
            
            fun(x) = -V(x,w,interp,par)       
            
            res = optimize(fun,1e-8,1-1e-8)
            
            sol.V[iw] = -res.minimum
            sol.C[iw] = res.minimizer*w
            

        end        
        sol.delta = findmax(abs.(sol.V .- V_next))[1]
    end 
    
    return sol
end


function V(x::Float64,w::Float64,interp::Interpolations.Extrapolation,par::NamedTuple)
    c = x*w

    wt1 = w-c

    return sqrt(c)+par.β *interp(wt1)
end

function solve_consumption_grid_search2(par::NamedTuple)

    
    sol = Sol(W_grid = par.W_grid,
            C= zeros(Float64,par.num_W),
            V= zeros(Float64,par.num_W),
            )
    
    grid_C = LinRange(0, 1.0, par.num_C)

    
    while (par.max_iter>= sol.it) & (par.tol<sol.delta)
        sol.it +=1
        V_next = copy(sol.V)

        
        interp = LinearInterpolation(par.W_grid,V_next)

        for (iw,w) in enumerate(par.W_grid)  # enumerate automaticcaly unpack w
            c = grid_C.*w
            wt1 = w .- c
            V_guess = sqrt.(c) .+ par.β .* interp.(wt1)
            
            star = findmax(V_guess)
            
            sol.V[iw] = star[1] 
            sol.C[iw] = (star[2]-1)*w
                        

        end        
        sol.delta = findmax(abs.(sol.V .- V_next))[1]
    end 
    
    return sol
end

end